app/scripts/controllers/network/network-controller.js

```diff
@@ -509,4 +525,69 @@ export default class NetworkController extends EventEmitter {
     this._provider = provider;
     this._blockTracker = blockTracker;
   }
+
+  /**
+   * Network Configuration management functions
+   */
+
+  /**
+   * Adds a network configuration if the rpcUrl is not already present on an
+   * existing network configuration. Otherwise updates the entry with the matching rpcUrl.
+   *
+   * @param config - The network configuration.
+   * @param config.rpcUrl - The network configuration RPC URL.
+   * @param config.chainId - The chain ID of the network, as per EIP-155.
+   * @param config.ticker - Currency ticker.
+   * @param config.chainName - Personalized network name.
+   * @param config.rpcPrefs - Personalized preferences.
+   * @returns uuid for the added or updated network configuration
+   */
+  upsertNetworkConfiguration({ rpcUrl, chainId, ticker, chainName, rpcPrefs }) {
+    const networkConfigurations = this.networkConfigurations.getState();
+    const newNetworkConfiguration = {
+      rpcUrl,
+      chainId,
+      ticker,
+      chainName,
+      rpcPrefs,
+    };
+
+    for (const [configUUID, networkConfiguration] of Object.entries(
+      networkConfigurations,
+    )) {
+      if (
+        networkConfiguration.rpcUrl?.toLowerCase() === rpcUrl?.toLowerCase()
+      ) {
+        networkConfigurations[configUUID] = newNetworkConfiguration;
+        this.networkConfigurations.updateState(networkConfigurations);
+        return configUUID;
+      }
+    }
+
+    const uuid = random();
+    networkConfigurations[uuid] = newNetworkConfiguration;
+    this.networkConfigurations.updateState(networkConfigurations);
+
+    return uuid;
+  }
+
+  /**
+   * Removes network configuration from state.
+   *
+   * @param uuid - Custom RPC URL.
+   */
+  removeNetworkConfiguration(uuid) {
+    const networkConfigs = this.networkConfigurations.getState();
```

Comment:
Nit: What do you think about being consistent about how we spell "network configuration" when it comes to naming variables in this file? Between `config`, `networkConfig`, `networkConfiguration`, my vote would be the last since it's the least ambiguous and is consistent with the name of the state property as well as the network configuration management methods, but I could go for the second if it's too long to type.

Reply:
Agreed. I've changed all of these instances to `networkConfiguration`